storage/backup/backup-cli/src/backup_types/transaction/backup.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    backup_types::transaction::manifest::{TransactionBackup, TransactionChunk},
+    storage::{BackupHandleRef, BackupStorage, FileHandle, ShellSafeName},
+    utils::{
+        backup_service_client::BackupServiceClient, read_record_bytes::ReadRecordBytes,
+        GlobalBackupOpt,
+    },
+};
+use anyhow::{ensure, Result};
+use libra_types::transaction::Version;
+use once_cell::sync::Lazy;
+use std::{convert::TryInto, mem::size_of, str::FromStr, sync::Arc};
+use structopt::StructOpt;
+use tokio::io::AsyncWriteExt;
+
+#[derive(StructOpt)]
+pub struct TransactionBackupOpt {
+    #[structopt(long = "start-version", help = "First transaction to backup.")]
+    pub start_version: u64,
+
+    #[structopt(long = "num_transactions", help = "Number of transactions to backup")]
+    pub num_transactions: usize,
+}
+
+pub struct TransactionBackupController {
+    start_version: u64,
+    num_transactions: usize,
+    max_chunk_size: usize,
+    client: Arc<BackupServiceClient>,
+    storage: Arc<dyn BackupStorage>,
+}
+
+impl TransactionBackupController {
+    pub fn new(
+        opt: TransactionBackupOpt,
+        global_opt: GlobalBackupOpt,
+        client: Arc<BackupServiceClient>,
+        storage: Arc<dyn BackupStorage>,
+    ) -> Self {
+        Self {
+            start_version: opt.start_version,
+            num_transactions: opt.num_transactions,
+            max_chunk_size: global_opt.max_chunk_size,
+            client,
+            storage,
+        }
+    }
+
+    pub async fn run(self) -> Result<FileHandle> {
+        let backup_handle = self.storage.create_backup(&self.backup_name()).await?;
+
+        let mut chunks = Vec::new();
+        let mut chunk_bytes = Vec::new();
+
+        let mut transactions_file = self
+            .client
+            .get_transactions(self.start_version, self.num_transactions)
+            .await?;
+        let mut current_ver: u64 = self.start_version;
+        let mut chunk_first_ver: u64 = self.start_version;
+
+        while let Some(record_bytes) = transactions_file.read_record_bytes().await? {
+            if chunk_bytes.len() + size_of::<u32>() + record_bytes.len() > self.max_chunk_size {
+                ensure!(
+                    !chunk_bytes.is_empty(),
+                    "max chunk size too small: {}",
+                    self.max_chunk_size
+                );
+                assert!(chunk_bytes.len() <= self.max_chunk_size);
```

Comment:
Would this fail if `max_chunk_size` is smaller than any record size?

Reply:
Yeah it dies. I was copying from the original state snapshot backup code. Converting to an `ensure!` is probably better.

Reply:
I think maybe it is better to limit the max_chunk_size to be `max(user_set_max_chunk_size, record_size)`.